### PR TITLE
feat: README and license (SC-3667)

### PR DESCRIPTION
**Pull request by JGcarv**
# Description

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,119 @@
...
+
+`ProxyFactory.sol`
+
+Responsible for deploying new instances and triggering initialization and migration logic atomically. 
+
+```js
+    contract ProxyFactory {
+
+        /// @dev Register a new implementation address attached to a version, which can be used with any scheme of versioning.
+        function _registerImplementation(uint256 version, address implementationAddress) internal virtual returns (bool success)
```
Comment (deluca-mike) [inline README.md:30]:
```suggestion
        function _registerImplementation(uint256 version, address implementationAddress) internal virtual returns (bool success);
```
and the rest as well.

**Review by deluca-mike — CHANGES_REQUESTED** [review]

**Review by deluca-mike — DISMISSED** [review]

File: README.md [hunk 2]
```diff
@@ -1 +1,119 @@
...
+        /// @dev Deploys a new instance of and calls the initialization function with provided arguments
+        function _newInstance(uint256 version, bytes calldata arguments) internal virtual returns (bool success, address proxy);
+
+        /// @dev Deploys a new instance at a specific address using the salt and calls the initialization function with provided arguments
+        function _newInstanceWithSalt(uint256 version, bytes calldata arguments, bytes32 salt) internal virtual returns (bool success, address proxy); 
+
+        /// @dev Calls the proxy with arguments to perform the necessary initialization
+        function _initializeInstance(address proxy, uint256 version, bytes calldata arguments) internal virtual returns (bool success); 
+
+        /// @dev Register a possible migration path and optionally sets a migrator contract
```
Comment (lucas-manuel) [inline README.md:41]: Register a migration path between versions and optionally set a migrator contract

Comment (lucas-manuel) [reply README.md:41]: Registers

File: README.md [hunk 3]
```diff
@@ -1 +1,119 @@
...
+### Contracts
+
+`ProxyFactory.sol`
+
+Responsible for deploying new instances and triggering initialization and migration logic atomically. 
+
+```js
+    contract ProxyFactory {
+
+        /// @dev Register a new implementation address attached to a version, which can be used with any scheme of versioning.
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply README.md:29]: Registers

File: README.md [hunk 4]
```diff
@@ -1 +1,119 @@
...
+        /// @dev Deploys a new instance at a specific address using the salt and calls the initialization function with provided arguments
+        function _newInstanceWithSalt(uint256 version, bytes calldata arguments, bytes32 salt) internal virtual returns (bool success, address proxy); 
+
+        /// @dev Calls the proxy with arguments to perform the necessary initialization
+        function _initializeInstance(address proxy, uint256 version, bytes calldata arguments) internal virtual returns (bool success); 
+
+        /// @dev Register a possible migration path and optionally sets a migrator contract
+        function _registerMigrationPath(uint256 fromVersion, uint256 toVersion, address migrator) internal virtual returns (bool success); 
+
+        /// @dev Updates the implementation used by a proxy.
```
Comment (lucas-manuel) [inline README.md:44]: Proxy

File: README.md [hunk 5]
```diff
@@ -1 +1,119 @@
...
+        function _upgradeInstance(address proxy, uint256 toVersion, bytes calldata arguments) internal virtual returns (bool success); 
+    }
+```
+
+`Proxy.sol`
+
+The slim contract that is deployed. It saves both and `implementation` and the `factory` address to be able to execute transactions and upgrades.
```
Comment (lucas-manuel) [inline README.md:51]: The Proxy contract that is deployed and manages storage.

Comment (lucas-manuel) [reply README.md:51]: addresses

File: README.md [hunk 6]
```diff
@@ -1 +1,119 @@
...
+
+```js
+contract Proxy is SlotManipulatable {
+
+    /// @dev Storage slot with the address of the current factory. This is the keccak-256 hash of "FACTORY_SLOT".
+    bytes32 private constant FACTORY_SLOT = 0xf2db84db8157f5a01a257d644038e8929d5a62c9ffa8b736374913908897e5bb;
+
+    /// @dev Storage slot with the address of the current factory. This is the keccak-256 hash of "IMPLEMENTATION_SLOT".
```
Comment (lucas-manuel) [inline README.md:59]: current implementation

File: README.md [hunk 7]
```diff
@@ -1 +1,119 @@
...
+    bytes32 private constant IMPLEMENTATION_SLOT = 0xf603533e14e17222e047634a2b3457fe346d27e294cedf9d21d74e5feea4a046;
+
+    /// @dev Function to be called right after deployment, similar to a constructor in regular contracts.
```
Comment (lucas-manuel) [inline README.md:62]: Function to be called right after deployment, to set the `implementation` and the `factory` addresses in storage.

File: README.md [hunk 8]
```diff
@@ -1 +1,119 @@
...
+    function _setup() private; 
+
+    /// @dev Function to delegatecall all incoming function to `implementation`
```
Comment (lucas-manuel) [inline README.md:65]: functions to the `implementation` address.

File: README.md [hunk 9]
```diff
@@ -1 +1,119 @@
...
+    fallback() payable external virtual; 
+
+}
+```
+
+`SlotManipulatable.sol`
+
+Helper contract that allow to manually modify storage that might be needed during a migration process
```
Comment (lucas-manuel) [inline README.md:73]: Helper contract that can manually modify storage when necessary (i.e., during a initialization/migration process)

File: README.md [hunk 10]
```diff
@@ -1 +1,119 @@
...
+
+ ```js
+ contract SlotManipulatable {
+
+    /// @dev returns the value stored at slot
```
Comment (lucas-manuel) [inline README.md:78]: Returns the value stored at the given slot.

File: README.md [hunk 11]
```diff
@@ -1 +1,119 @@
...
+    function _getSlotValue(bytes32 slot) internal view returns (bytes32 value); 
+
+    /// @dev set the storage slot to the given value
```
Comment (lucas-manuel) [inline README.md:81]: Sets the value stored at the given slot.

File: README.md [hunk 12]
```diff
@@ -1 +1,119 @@
...
+    function _setSlotValue(bytes32 slot, bytes32 value) internal; 
+
+    // @dev Get the storage slot for a reference type
```
Comment (lucas-manuel) [inline README.md:84]: Returns the storage slot for a reference type.

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Still need to review the bottom, but here's the first sweep

File: README.md (hunk 2 above)
Comment (JGcarv) [reply README.md:41]: What's the correct form?

Register a migration path between versions and optionally set a migrator contract. 
or
Register a migration path between versions and optionally _sets_ a migrator contract

File: README.md [hunk 13]
```diff
@@ -1 +1,119 @@
...
+
+- **Migration Contracts:** Architecture allows for intermediary contracts that can perform storage migration operations between two versions on upgrade, as well as perform initialize functionality on Proxy instantiation.
+
+### Contracts
+
+`ProxyFactory.sol`
+
+Responsible for deploying new Proxy instances and triggering initialization and migration logic atomically. 
+
+```js
```
Comment (lucas-manuel) [inline README.md:26]: We almost shouldnt use `js` here since the highlighting doesn't seem to be working.

File: README.md [hunk 14]
```diff
@@ -1 +1,119 @@
...
+    contract ProxyFactory {
+
+        /// @dev Registers a new implementation address attached to a version, which can be used with any uint256 versioning scheme.
+        function _registerImplementation(uint256 version, address implementationAddress) internal virtual returns (bool success);
+
+        /// @dev Deploys a new Proxy instance of and calls the initialization function with provided arguments.
```
Comment (lucas-manuel) [inline README.md:32]: instance and

File: README.md [hunk 15]
```diff
@@ -1 +1,119 @@
...
+        function _newInstance(uint256 version, bytes calldata arguments) internal virtual returns (bool success, address proxy);
+
+        /// @dev Deploys a new new Proxy instance at a specific address using a salt and calls the initialization function with provided arguments.
+        function _newInstanceWithSalt(uint256 version, bytes calldata arguments, bytes32 salt) internal virtual returns (bool success, address proxy); 
+
+        /// @dev Calls the Proxy with arguments to perform the necessary initialization.
+        function _initializeInstance(address proxy, uint256 version, bytes calldata arguments) internal virtual returns (bool success); 
+
+        /// @dev Registers a migration path between versions and optionally set a migrator contract
```
Comment (lucas-manuel) [inline README.md:41]: sets

Comment (lucas-manuel) [reply README.md:41]: Add period

File: README.md [hunk 16]
```diff
@@ -1 +1,119 @@
...
+## Testing and Development
+#### Setup
+```sh
+git clone git@github.com:maple-labs/maple-core.git
+cd maple-core
+dapp update
+```
+#### Running Tests
+- To run all tests: `make test` (runs `./test.sh`)
+- To run a specific test function: `./test.sh <test_name>` (e.g. `./test.sh test_composability`)
```
Comment (lucas-manuel) [inline README.md:99]: `./test.sh -t <test_name>` (e.g. `./test.sh -t test_composability`)

File: README.md [hunk 17]
```diff
@@ -1 +1,119 @@
-## proxy-factory
+# Proxy Factory
+
+[![CircleCI](https://circleci.com/gh/maple-labs/proxy-factory/tree/main.svg?style=svg)](https://circleci.com/gh/maple-labs/proxy-factory/tree/main) [![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)
+
+
+Set of base contracts to deploy and manage versions on chain, designed to be minimally opinionated, extensible and gas-efficient. These contracts were built to provide the necessary features to be reused across multiple projects, both within Maple and externally.
+
+While there are other good libraries for base smart contracts, none of them satisfied the need to simultaneously be feature complete and allow to be used within our own style of development.
+
```
Comment (lucas-manuel) [inline README.md:9]: Single new line

File: README.md [hunk 18]
```diff
@@ -1 +1,119 @@
...
+git clone git@github.com:maple-labs/maple-core.git
+cd maple-core
+dapp update
+```
+#### Running Tests
+- To run all tests: `make test` (runs `./test.sh`)
+- To run a specific test function: `./test.sh <test_name>` (e.g. `./test.sh test_composability`)
+
+This project was built using <a href="https://github.com/dapphub/dapptools">dapptools</a>
+
```
Comment (lucas-manuel) [inline README.md:102]: Single new line

File: README.md [hunk 19]
```diff
@@ -1 +1,119 @@
...
+
+
+## Security
+
+The code is designed to be extended with needed feature, including access control and other guards on deployment and upgradeability, therefore it's not advised to use this as is. Additionally, it is still in development and haven't been externally audited.
```
Comment (lucas-manuel) [inline README.md:107]: The code is designed to be highly flexible and extensible, meaning that logic that is usually part of these functions  (e.g., access controls) was not included. Therefore **it is strongly advised that these contracts be implemented with proper sanity checks, access controls and any extra logic necessary for security.**

Comment (lucas-manuel) [reply README.md:107]: Lets also add our boilerplate disclaimer at the top of this README.

File: README.md [hunk 20]
```diff
@@ -1 +1,119 @@
...
+
+This project was built using <a href="https://github.com/dapphub/dapptools">dapptools</a>
+
+
+
+## Security
+
+The code is designed to be extended with needed feature, including access control and other guards on deployment and upgradeability, therefore it's not advised to use this as is. Additionally, it is still in development and haven't been externally audited. 
+
+
```
Comment (lucas-manuel) [inline README.md:109]: Single newline

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: README.md [hunk 21]
```diff
@@ -1 +1,119 @@
-## proxy-factory
+# Proxy Factory
+
+[![CircleCI](https://circleci.com/gh/maple-labs/proxy-factory/tree/main.svg?style=svg)](https://circleci.com/gh/maple-labs/proxy-factory/tree/main) [![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)
+
+
+Set of base contracts to deploy and manage versions on chain, designed to be minimally opinionated, extensible and gas-efficient. These contracts were built to provide the necessary features to be reused across multiple projects, both within Maple and externally.
+
+While there are other good libraries for base smart contracts, none of them satisfied the need to simultaneously be feature complete and allow to be used within our own style of development.
```
[the remark this answers is not in the document]
Comment (JGcarv) [reply README.md:8]: Which one? `While...development?`

File: README.md (hunk 13 above)
Comment (JGcarv) [reply README.md:26]: I can remove, but I personally thinks that with no syntax highlighting at all it becomes harder to read.

File: README.md (hunk 19 above)
Comment (JGcarv) [reply README.md:107]: Where can I find this disclaimer?

**Pull request merged**